### PR TITLE
Add Jetson Xavier AGX/NX SD/NX eMMC balenaOS migration support 

### PR DESCRIPTION
**Pull request by acostach**
Add balenaOS to balenaOS migration support for Jetson Xavier AGX, NX SD and NX eMMC devices
    
In this PR we:

      - ensure the existing config.json is preserved during the migration
      - all connection files in /mnt/boot/system-connections are transferred from the old OS
      - all system-proxy files in /mnt/boot/system-proxy are transferred as well
      - the eMMC or SD-card of the above-mentioned devices is flashed with the new OS
      - the QSPI or eMMC boot partition is flashed with the boot blob from the new OS
    
Signed-off-by: Alexandru Costache <alexandru@balena.io>

Tested on AGX Xavier with

```
mkdir migrate
PATH="${PATH}:/mnt/data/migrate/bin" ./takeover -i balena-image-jetson-xavier.balenaos-img.gz --no-nwmgr-check --log-file /mnt/data/migrate/stage1.log --log-level trace -l /dev/sda1 --s2-log-level trace
```

### CODE REVIEW

File: src/common/options.rs [hunk 1]
```diff
@@ -25,6 +25,13 @@ pub struct Options {
         help = "Path to balena-os image"
     )]
     image: Option<PathBuf>,
+    #[clap(
+        long = "ldd-path",
+        value_name = "LDD_PATH",
+        value_parser,
+        help = "Path to ldd script"
+    )]
+    ldd_path: Option<PathBuf>,
```
Comment (lmbarros) [inline src/common/options.rs:34]: That's one of these points I think we'll need to discuss in wider audience, but I am personally not a fan of this solution (I chatted a bit with @rahul-thakoor about this, I think he agrees with me). The reasoning is:

1. All our motivation for this work is to enable balenaOS-to-balenaOS migrations.
2. balenaOS does not provide an `ldd`.
3. So, with this solution we are just moving the friction away from us and towards our target users.

I think it would make much more customer-friendly that we solved this ourselves, either in takeover or in the OS.

**Review by lmbarros — COMMENTED** [review]
@acostach , I am leaving just some initial comments that don't really get much into the main point of this PR. I'll need more time to get the context I need to make a proper review. Planning to keep looking into this next week!

File: src/common/options.rs (hunk 1 above)
Comment (rahul-thakoor) [reply src/common/options.rs:34]: Yes, i'm not convinced we can ask end users to provide their own `ldd` script. I think
- we either ship it in balenaOS and backport where required. customers have to HUP to version that has `ldd` then HUP to next major version
- we find alternatives and do it natively in the rust code. i looked at it briefly a while back, but https://crates.io/crates/lddtree looked promising but would require some changes

File: src/stage1.rs [hunk 2]
```diff
@@ -272,7 +324,17 @@ fn prepare(opts: &Options, mig_info: &mut MigrateInfo) -> Result<()> {
 
     // *********************************************************
     // make mountpoint for tmpfs
-    let takeover_dir = PathBuf::from(TAKEOVER_DIR);
+    let mut takeover_dir = PathBuf::from("");
+    if mig_info.os_name().starts_with(BALENA_OS_NAME) {
+        // base directory for mountpoint must be writeable
+        takeover_dir.push(BALENA_DATA_MP);
+    }
```
[the remark this answers is not in the document]
Comment (acostach) [reply src/stage1.rs:331]: My tests have been focuses on on the Jetsons, which have a data and a boot partition by default, just like any other balenaOS device, but are you thinking of creating a new tmpfs or using a path inside /tmp?

File: src/stage2.rs [hunk 3]
```diff
@@ -423,101 +431,172 @@ fn transfer_boot_files<P: AsRef<Path>>(dev_root: P) -> Result<()> {
...
+                    boot_part = Some(partition);
+                }
+                2 => {
+                    root_a_part = Some(partition);
+                }
+                3..=5 => debug!("Skipping partition {}", partition.index),
+                6 => {
+                    data_part = Some(partition);
+                    break;
+                }
```
[the remark this answers is not in the document]
Comment (acostach) [reply src/stage2.rs:530]: Good point @lmbarros , would a comment with some explanations and a check to bail-out if the source OS is not balenaOS be sufficient, or should we add something else?

File: src/common/options.rs (hunk 1 above)
Comment (acostach) [reply src/common/options.rs:34]: @lmbarros @rahul-thakoor hmm, in the context of cloud triggered migrations, an approach like this would allow the action script to place the `ldd` script near the `takeover` binary. So I don't think this moves the friction to the target users, because they won't be starting takeover. In the case of a balenaOS to balenaOS migration, that's handled by the action script.

The next best option, apart from the action script handling ldd, would be takeover to either include the script or implement it like you said. But that would be a different effort, which would need to be also maintained. Upstream changes to `ldd` may need to be monitored and maintained, even if they aren't all that frequent.

I'd like to avoid having to tie migration to ldd presence in the hostOS, because that would prevent migration for any device that's not updated to the latest release, which adds friction by requiring an extra step of hostOS update for users.

File: src/common/system.rs [hunk 4]
```diff
@@ -324,40 +324,46 @@ pub(crate) fn fuser<P: AsRef<Path>>(
...
                                         }
-                                        break;
                                     }
-                                } else {
-                                    return Err(Error::with_context(
-                                        ErrorKind::InvState,
-                                        &format!("file '{}' is not a link", curr_path.display()),
-                                    ));
+                                    Err(_) => {
+                                        debug!("continue after lstat error");
```
Comment (lmbarros) [inline src/common/system.rs:366]: This whole set of changes in the `fuser()` function seems to be "just" to allow continuing after an `lstat()` error. Do you know why this is necessary with balenaOS-to-balenaOS migrations but wasn't necessary before? We should document the reason.

File: src/stage1.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (lmbarros) [reply src/stage1.rs:331]: I don't have any specific idea in mind... My comment was really very generic. In summary: It would be better *if* we could find a way to avoid treating balenaOS as a special case.

Each different code path we create to handle a special makes testing and maintenance in general a bit more difficult.

My understanding is that when Ken made these changes he was just trying to get the balenaOS-to-balenaOS migration working as proof-of-concept. So, I believe at that point he (rightfully) wasn't worried by this kind of longer-term issues, but now I think bow we should start to worry about this.

Again, I didn't give any though to this issue myself yet, but I will get there this week!

File: src/stage2.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (lmbarros) [reply src/stage2.rs:530]: Frankly, I think we can just ignore this for now @acostach . There are plenty of things we could do to improve the code base, but for this PR I believe it's wiser to focus on the balenaOS-to-balenaOS changes.

(I know, I was the one to start this discussion. Shame on me!)

File: src/common/options.rs (hunk 1 above)
Comment (lmbarros) [reply src/common/options.rs:34]: > In the context of cloud triggered migrations, an approach like this would allow the action script to place the ldd script near the takeover binary.

That's a good point, @acostach , which I honestly wasn't keeping in mind! Still, here's a possible counterpoint: I was under the impression that we'll have a period in which users will run takeover manually for balenaOS-to-balenaOS migrations. During this period, the friction would be on their side.

And of course you are right that it would be an additional effort for us to somehow provide `ldd`. We need to weigh this in, too.

> I'd like to avoid having to tie migration to ldd presence in the hostOS, because that would prevent migration for any device that's not updated to the latest release, which adds friction by requiring an extra step of hostOS update for users.

Yup, I get this and it makes sense! I think that would be my current favorite approach, too.

But, food for thought: requiring this extra intermediate HUP could also potentially be a bit safer, because it would force all balenaOS-to-balenaOS migrations to be from an OS release that is at least close to what we tested. Given that this is a risky, not-undoable operation, I sometimes wonder if this extra HUP wouldn't be a good idea.

File: src/stage1.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (lmbarros) [reply src/stage1.rs:331]: OK, I did some tests here and I believe that we could simply define `TAKEOVER_DIR` like this in `defs.rs`:

```
pub(crate) const TAKEOVER_DIR: &str = "/tmp/balena-takeover";
```

and then get rid of all the special handling that we added in order to handle balenaOS. Here's how I tested:

* I changed `master` to use `/tmp/balena-takeover`.
* Then I used this version to migrate a Pi 3 from RaspiOS to balenaOS.
* Then I flashed a new RaspiOS, changed its `/tmp` to be a `tmpfs` like in balenaOS and tried to migrate. It worked.

@acostach , do you have a simple way to try using  `/tmp/balena-takeover` on a Jetson migration?

If this works, we can remove a lot of the code we added.

File: Cargo.toml [hunk 5]
```diff
@@ -8,11 +8,17 @@ publish = false
...
+opt-level = "s"
+
 [dependencies]
 
 [dependencies.libc]
 version = "0.2.70"
 
+[dependencies.finder]
+version = "0.1"
+
```
Comment (lmbarros) [inline Cargo.toml:21]: Your call here, but I'd suggest to:

1. Make the size optimization a separate commit.
2. Move the "add finder dependency" to the same commit in which you use this `finder` package.

I think this would make each commit more self-contained, and therefore:

1. Easier to understand if we ever need to do Git archaeology in the future.
2. Easier to revert, shall we find some issue.

File: src/common/defs.rs [hunk 6]
```diff
@@ -11,6 +11,11 @@ pub(crate) const BLKID_CMD: &str = "blkid";
 pub(crate) const EFIBOOTMGR_CMD: &str = "efibootmgr";
 pub(crate) const DD_CMD: &str = "dd";
 
+/* The mdtd_debug tool is used on Xavier NX devices to clear and write the QSPI
+   with the blob included in the target OS image
+*/
+pub(crate) const MTD_DEBUG_CMD: &str = "mtd_debug";
```
Comment (lmbarros) [inline src/common/defs.rs:17]: This is another of my more generic suggestions. One problem I think takeover suffers from already is assuming that the future maintainers know everything the original developer knew.  I often find hard to understand what exactly some part of the code is doing, and especially why.

And here's one case in which I think we are making this situation worse. I personally don't know what `mtd_debug` is, and a quick search gives me only a vague idea of what it is. So, I think a few lines of comments here could be very helpful in the long run. A link to wherever this tool is documented could also help a lot.

File: src/common/disk_util.rs [hunk 7]
```diff
@@ -286,7 +286,7 @@ impl<'a> PartitionIterator<'a> {
                     self.part_idx = 4;
                     self.get_extended_partition()
                 }
-                PartitionType::Fat | PartitionType::Linux => {
+                PartitionType::Fat | PartitionType::Linux | PartitionType::GPT => {
```
Comment (lmbarros) [inline src/common/disk_util.rs:289]: Why this wasn't needed before? Is it because previously we'd only iterate over partitions of balenaOS versions using MBR partitions? Please consider explaining the "why" of this change in the commit message.

File: src/stage1/exe_copy.rs [hunk 8]
```diff
@@ -50,7 +54,14 @@ impl ExeCopy {
...
         trace!("get_libs_for: entered");
-        let ldd_path = whereis("ldd").upstream_with_context("Failed to locate ldd executable")?;
+        let ldd_path: String;
+        if self.ldd_script_path.is_empty() {
+            debug!("lld path was not provided, will use the one from current OS, if available");
+            ldd_path = whereis("ldd").upstream_with_context("Failed to locate ldd executable, please provide path to ldd manually")?;
+        } else {
+            debug!("Provided ldd path: {}", self.ldd_script_path.display());
+            ldd_path = self.ldd_script_path.as_path().display().to_string();
+        }
```
Comment (lmbarros) [inline src/stage1/exe_copy.rs:64]: Like in a case I commented earlier, I think it would make sense to keep the inclusion and the usage of the ldd-path option on a single commit.

File: src/stage1/migrate_info.rs [hunk 9]
```diff
@@ -43,18 +43,26 @@ pub(crate) struct MigrateInfo {
...
+        let os_name = get_os_name()?;
+        info!("Detected device type: {} running {}", device.get_device_type(), os_name);
 
+        /* If no config.json is passed in command line and we're running on balenaOS,
+         * we can preserve the existing config.json
+         */
         let mut config = if let Some(balena_cfg) = opts.config() {
             BalenaCfgJson::new(balena_cfg)?
-        } else {
+        } else if Path::new("/mnt/boot/config.json").exists() {
```
Comment (lmbarros) [inline src/stage1/migrate_info.rs:62]: I believe you intended to check also if  `os_name == "balenaOS"` or something like that, given that you added all the infrastructure to read the OS name!

File: src/common/system.rs [hunk 10]
```diff
@@ -391,7 +392,7 @@ pub(crate) fn fuser<P: AsRef<Path>>(
         sleep(if let Some(wait_for_term) = wait_for_term {
             wait_for_term
         } else {
-            Duration::from_millis(500)
+            Duration::from_millis(3000)
```
Comment (lmbarros) [inline src/common/system.rs:395]: Hey @acostach , did you notice any other warnings after you changed this?

File: src/stage1.rs [hunk 11]
```diff
@@ -76,18 +75,65 @@ fn prepare_configs<P1: AsRef<Path>>(
...
+        for path in nwmgr_files {
+            mig_info.add_nwmgr_file(path.unwrap().path());
+        }
+
+        debug!("migrating from balenaOS - copying system-proxy files");
+        let system_proxy_files = read_dir(BALENA_SYSTEM_PROXY_BOOT_PATH).unwrap();
+        for sys_proxy_path in system_proxy_files {
+            mig_info.add_system_proxy_file(sys_proxy_path.unwrap().path());
+        }
+    }
```
Comment (lmbarros) [inline src/stage1.rs:107]: It makes sense to copy all connections in a balenaOS-to-balenaOS migration... but we already some code related with migration of network connections. I don't know if that existing code can interfere with this new one in any way, but we should check this. In any case, I'd prefer to try to find a way to keep all this stuff together.

The code I am referring to lives in `wifi_configs.rs` and seems to be specific to Wifi connections, but still related.

File: src/stage1/migrate_info.rs [hunk 12]
```diff
@@ -202,10 +213,22 @@ impl MigrateInfo {
...
         self.device.supports_device_type(DEV_TYPE_GEN_X86_64)
     }
 
+    pub fn is_jetson_xavier(&self) -> bool {
+        self.device.supports_device_type(DEV_TYPE_JETSON_XAVIER)
+    }
+
+    pub fn is_jetson_xavier_nx(&self) -> bool {
+        self.device.supports_device_type(DEV_TYPE_JETSON_XAVIER_NX)
+    }
```
[the remark this answers is not in the document]
Comment (lmbarros) [reply src/stage1/migrate_info.rs:230]: Well, your code is really just doing what `is_x86()` already did. It was [my commit that introduced `--no-dt-check`](https://github.com/balena-os/takeover/pull/54/commits/91e731c3c2e333d781d50c715469da8fde683b2b) that break things. I didn't anticipate this interaction between `--no-dt-check` and the points in the code where we check for certain device types (or "device families").

I am always happy to hear suggestions, but feel free to focus on my other comments. This point here is on me!

File: src/common/system.rs (hunk 4 above)
Comment (kb2ma) [reply src/common/system.rs:366]: @lmbarros, @acostach See the Fibery page _Experiment: Takeover MVP_, section _Stage 2_. On the NUC conversion, takeover was unable to find a file in the proc filesystem. I could not find it manually either, so I figured it was not a significant issue. I never got to the level of understanding _why_ the file was missing. So this change essentially just ignores the error. The Fibery page has more ideas on the cause. Have either of you seen this error in the logs during a migration?

File: src/common/system.rs (hunk 10 above)
Comment (acostach) [reply src/common/system.rs:395]: @lmbarros after following your suggestion and increasing the timeout here I did not see any more warning on balena-engine still running

File: src/common/system.rs (hunk 4 above)
Comment (lmbarros) [reply src/common/system.rs:366]: Thanks, @kb2ma ! I have seen this error a couple of times. I actually moved the [printing of the process table](https://github.com/balena-os/takeover/pull/54/commits/513810c7b31cd865eed908b096ae79b43717ce2c) to a point before the error happens to help with troubleshooting it. (But after that I didn't look further.)

My best guess was that this was a racing issue: if a process exits between the moments we:

1. call `read_dir()` to get the list of entries under `/proc`
2. try to access that process' `/proc/PID/fd/...`

then we'll get this error.

I don't know if this is the only possible way to get this error, but this looks like one possible way at least. (I might be missing something, too... this code is tricky.)

Not very sure what to do here, @acostach , but "by default" I prefer to not merge fixes until we reasonably certain of what the issue really is.

File: src/stage1/migrate_info.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (acostach) [reply src/stage1/migrate_info.rs:230]: What where the errors on the Pi3? Was it related to what was reported in /proc/device-tree/model ?

File: src/stage1.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (acostach) [reply src/stage1.rs:331]: Unfortunately this doesn't work, the device reboots right after telinit is called and no stage2 logs are saved. Full discussion in https://balena.zulipchat.com/#narrow/stream/345889-balena-io.2Fos/topic/balenaOS.20to.20balenaOS.20migrator/near/428129684

File: src/stage2.rs [hunk 13]
```diff
@@ -653,7 +675,52 @@ fn efi_setup(device: &Path) -> Result<()> {
...
 
-fn raw_mount_balena(device: &Path) -> Result<()> {
+fn write_boot_blob(s2_config: &Stage2Config, mount_path: PathBuf)
+{
+    let file_finder = Finder::new(mount_path.clone());
+    let mut img_path = PathBuf::new();
+
+
+    if s2_config.device_type.starts_with("Jetson Xavier AGX") {
+        for i in file_finder.into_iter() {
```
[the remark this answers is not in the document]
Comment (acostach) [reply src/stage2.rs]: Addressed in https://github.com/balena-os/takeover/pull/66/commits/80bbbd4d7ee96eb94b29c1d50b77f9f9904c0226#diff-e24518d69b6357926bce35868225330c1992a4029b521c0112b4ec8523ddc31cR79

File: src/common/defs.rs [hunk 14]
```diff
@@ -11,6 +11,11 @@ pub(crate) const BLKID_CMD: &str = "blkid";
 pub(crate) const EFIBOOTMGR_CMD: &str = "efibootmgr";
 pub(crate) const DD_CMD: &str = "dd";
 
+/* The mdtd_debug tool is used on Xavier NX devices to clear and write the QSPI
+   with the blob included in the target OS image
+*/
```
[the remark this answers is not in the document]
Comment (acostach) [reply src/common/defs.rs]: @lmbarros I addressed this in https://github.com/balena-os/takeover/pull/66/commits/038b56c8215747041ec31ee7a6fce45b852f6f97

File: src/stage2.rs [hunk 15]
```diff
@@ -423,101 +431,172 @@ fn transfer_boot_files<P: AsRef<Path>>(dev_root: P) -> Result<()> {
...
+        for (i, p) in gpt.iter() {
+            if p.is_used() {
+                debug!("Partition #{}: type = {:?}, size = {} bytes, starting lba = {}, name = {}",
+                    i,
+                    p.partition_type_guid,
+                    p.size().unwrap() * gpt.sector_size,
+                    p.starting_lba,
+                    p.partition_name.as_str());
+
+                if  p.partition_name.as_str() == "resin-boot" {
```
[the remark this answers is not in the document]
Comment (acostach) [reply src/stage2.rs]: Addressed in https://github.com/balena-os/takeover/pull/66/commits/09e0df98384d364402ecd64e088cd2c7bfac9ee7#diff-44997d32ba1fbd5988c70c5522a6a3ea8612a1d161c2313fa0f7156c808b92d9R568

File: src/stage2.rs [hunk 16]
```diff
@@ -423,101 +431,172 @@ fn transfer_boot_files<P: AsRef<Path>>(dev_root: P) -> Result<()> {
...
+                } else if  p.partition_name.as_str() == BALENA_ROOTA_PART {
+                            root_a_part = Some(PartInfo {
+                                index: i as usize,
+                                ptype: 0x83,  // MBR Linux byte; really this is the EFI system
+                                              // partition, but MBR doesn't define this type.
+                                status: 0,    // Not clear what this should be
+                                start_lba: p.starting_lba,
+                                num_sectors: p.size().unwrap()
+                            })
+                } else if p.partition_name.as_str() == "resin-data" {
```
[the remark this answers is not in the document]
Comment (acostach) [reply src/stage2.rs]: Again thanks for spotting, fixed in https://github.com/balena-os/takeover/commit/09e0df98384d364402ecd64e088cd2c7bfac9ee7#diff-44997d32ba1fbd5988c70c5522a6a3ea8612a1d161c2313fa0f7156c808b92d9R568

File: Cargo.toml (hunk 5 above)
Comment (acostach) [reply Cargo.toml:21]: Sure,
1 is fixed in https://github.com/balena-os/takeover/pull/66/commits/1f8f596b8178e0bea2fb4b86f8faed0237244d3f
and 2 is addressed in https://github.com/balena-os/takeover/pull/66/commits/09e0df98384d364402ecd64e088cd2c7bfac9ee7

Was accustomed to committing changes per file/directory, as we usually do in device integration repos :)

File: src/common/defs.rs (hunk 6 above)
Comment (acostach) [reply src/common/defs.rs:18]: Sure, I added more details on mtd_debug and the mtd-utils package in https://github.com/balena-os/takeover/pull/66/commits/038b56c8215747041ec31ee7a6fce45b852f6f97

File: src/common/defs.rs [hunk 17]
```diff
@@ -26,14 +31,39 @@ pub const DISK_BY_LABEL_PATH: &str = "/dev/disk/by-label";
...
+pub const SYSTEM_PROXY_DIR: &str = "system-proxy";
+
+pub const BALENA_DATA_MP: &str = "/mnt/data/";
+pub const BALENA_OS_NAME: &str = "balenaOS";
+
+pub const BALENA_SYSTEM_CONNECTIONS_BOOT_PATH: &str = "/mnt/boot/system-connections/";
+pub const BALENA_SYSTEM_PROXY_BOOT_PATH: &str = "/mnt/boot/system-proxy/";
+
+/* Enables writing to the hardware-defined boot partition on AGX Xavier */
+pub const JETSON_XAVIER_HW_PART_FORCE_RO_FILE: &str = "/sys/block/mmcblk0boot0/force_ro";
```
[the remark this answers is not in the document]
Comment (acostach) [reply src/common/defs.rs:71]: Absolutely, more details on Linux boot partitions added in https://github.com/balena-os/takeover/pull/66/commits/038b56c8215747041ec31ee7a6fce45b852f6f97

File: src/common/disk_util.rs (hunk 7 above)
Comment (acostach) [reply src/common/disk_util.rs:289]: Indeed, updated in https://github.com/balena-os/takeover/pull/66/commits/3f5f2dab381e922a441b7ad61de480a4e69a191d

File: src/stage1/exe_copy.rs (hunk 8 above)
Comment (acostach) [reply src/stage1/exe_copy.rs:64]: Sure, added in the same commit in https://github.com/balena-os/takeover/pull/66/commits/4bbfe8aa7849ff33f88ee6951fe93c4999633533

File: src/stage1/migrate_info.rs (hunk 9 above)
Comment (acostach) [reply src/stage1/migrate_info.rs]: Addressed in https://github.com/balena-os/takeover/pull/66/commits/2df5e8c7f21d74732fe1ad71a09e06f2d15dec17#diff-fc1ecdab5d089a49201634bf1ca9669e4f4289a4fbc7f8ac77b144ab883050a1R62

File: src/stage1.rs (hunk 11 above)
Comment (acostach) [reply src/stage1.rs:106]: Hmm, we need to discuss this in the call. `wifi_config.rs`, `nwmgr_parser.rs` seem to be meant to for different types of functionality, like parsing NM/WPA Supplicant/Connman connection files, creating new ones, etc. We may want to disable that entirely for balenaOS to balenaOS migrations.

What I did in this part is to add the functionality of copying the existing system-proxy and system-connections to the tmpfs, in the case of migrating from balenaOS. This was and still being done for the connection files generated by `wifi_config.rs`, in the same place here in `stage1.rs.`

What I added doesn't interfere with the existing functionality, which still creates NM connection files apart from the ones I copy. The current way to disable creation of new NM connections by takeover is by using the existing '--no-nwmgr-check' flag.

File: src/stage1/migrate_info.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (lmbarros) [reply src/stage1/migrate_info.rs:228]: > Was it related to what was reported in /proc/device-tree/model ?

Yes, this. If I recall correctly, my Pi 3 reports itself as a "3 Model B Plus Rev 1", and our regex couldn't handle both "Plus" and "Rev". Or something like that.

Some more context: We decided to go with these `--no-*-check` flag instead of adjusting the regex (and similar changes for any other DTs and OSes) because this path would require continuous maintenance of `takeover` as the ecosystem evolves, and for the foreseeable future we'd rather spend this effort in other tasks.

[the remark this answers is not in the document]
Comment (acostach) [reply src/stage1/migrate_info.rs:228]: Wondering if we should care if it's a Pi3 B or B+, with or without a rev, because it should be able to run both the 32bit and the 64bit balenaOS images. Looking at https://www.seeedstudio.com/blog/2019/11/29/raspberry-pi-3-b-comparison-with-3b-and-more/ I see mostly networking improvements, nothing that prevent it from booting...

File: src/common/defs.rs (hunk 6 above)
Comment (acostach) [reply src/common/defs.rs:18]: Anything else I should add here?

File: src/common/disk_util.rs (hunk 7 above)
Comment (acostach) [reply src/common/disk_util.rs:289]: Fine to mark this one as resolved or should I add more to the commit message/comments?

File: src/stage1.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (acostach) [reply src/stage1.rs]: Turns out both init.rs and stage1.rs needed changes to use /tmp/balena-takeover as root mount-point, as in https://github.com/balena-os/takeover/pull/66/commits/a64fd6f0940d2694d416100c17616fd6656acb08

File: src/common/disk_util.rs (hunk 7 above)
Comment (lmbarros) [reply src/common/disk_util.rs:289]: It's fine, thanks!

File: src/stage1/migrate_info.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (lmbarros) [reply src/stage1/migrate_info.rs:234]: @acostach , I was experimenting a bit, and came up with [this commit on a test branch](https://github.com/balena-os/takeover/commit/8ea4be5d4a4624884016e6684b86bcf94f95c629). It makes sure that no DT-specific code runs when we pass `--no-dt-check`.

I tested this commit on top of your branch and migrated my Pi 3 (from Raspberry Pi OS) passing `--no-dt-check` without problems.

What do you think? Feel free to cherry-pick, adapt or steal as you see fit!

(Rant: The whole handling of DTs on Takeover feels a bit over-complicated to me. There's so much code around the `Device` trait and all its implementations, yet they seem provide so little functionality... they don't even encapsulate the device-specific code we have :disappointed: End of rant. And this was just a rant, changing this wouldn't make sense at this point, IMO.)

File: src/stage1.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (lmbarros) [reply src/stage1.rs]: So, I tried your branch on a Pi 3 and it worked fine, so looks like we can use `/tmp` as the mount point with the latest changes you made! :tada:  I think we can mark this concersation as resolved, right?

File: src/common/options.rs (hunk 1 above)
Comment (lmbarros) [reply src/common/options.rs:34]: I vote to mark this one as resolved, too. We'll need to revisit the `ldd` issue, but IMO this is not a blocker to the PR.

File: src/common/defs.rs (hunk 6 above)
Comment (lmbarros) [reply src/common/defs.rs:18]: Nope! That was a great improvement, thanks!

File: src/common/system.rs (hunk 4 above)
Comment (acostach) [reply src/common/system.rs:369]: ack, marking as resolved

File: src/common/options.rs (hunk 1 above)
Comment (acostach) [reply src/common/options.rs:34]: sure, marking as resolved

File: src/stage1.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (acostach) [reply src/stage1.rs]: Indeed, /tmp/ works for me too for balenaOS to balenaOS on both NUC and Jetson AGX. marking as solved

File: src/stage1/migrate_info.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (acostach) [reply src/stage1/migrate_info.rs:234]: @lmbarros lgtm. Should I bring your commit in this PR?

[the remark this answers is not in the document]
Comment (lmbarros) [reply src/stage1/migrate_info.rs:234]: I think so @acostach !

[the remark this answers is not in the document]
Comment (acostach) [reply src/stage1/migrate_info.rs:234]: thanks @lmbarros, done!

File: src/stage1.rs (hunk 11 above)
Comment (lmbarros) [reply src/stage1.rs]: > What I added doesn't interfere with the existing functionality, which still creates NM connection files apart from the ones I copy. The current way to disable creation of new NM connections by takeover is by using the existing '--no-nwmgr-check' flag.

Yes, I get this @acostach ! To be clear, my concern is not about this being incorrect, it's all about trying to make the code as whole more consistent to make longer-term easier. And also: I don't think the existing code was particularly to follow (maybe it's just me?), so I don't expect us to do any miracles here -- but I am still hopeful we can make it a bit more consistent.

I created this [commit sketching of a suggestion](https://github.com/balena-os/takeover/commit/bbe6fd67652f9f3e1ed7f459907f7b9caf24eabf). This one is not meant to be cherry-picked! It's not tested. I created the commit just because I think it's easier than trying to explain in words. But the idea was to try to make the new code more aligned with the code we already had for handling WiFi networks:

1. In `migrate_info.rs` populate the `MigrateInfo` fields. [Like we did before](https://github.com/balena-os/takeover/blob/f2ed41a7703db195e5427dcfe1a184e5820b5295/src/stage1/migrate_info.rs#L127) for `wifis`.
2. In `stage1.rs` do the actual copying. Again, this is similar to what we did for `wifis`: [this is the same place](https://github.com/balena-os/takeover/blob/f2ed41a7703db195e5427dcfe1a184e5820b5295/src/stage1.rs#L103
) we currently already create the connection files for WiFi.

(As a bonus suggestion, I also tossed in some smaller changes to propagate the errors up the call stack instead of using `unwrap()`, which panics in case of errors. Just for more friendly error reporting.)

In summary, what I am proposing is not a big change, but I think it helps to keep similar operations closer together in the code, which hopefully will help future maintainers to understand it.

Comment (lmbarros) [reply src/stage1.rs]: @acostach , I was just now reviewing our Project on Fibery. Sorry, I didn't recall we had a separate task to deal with this "network migration" point.

From my perspective, it would be OK to merge this PR as it is right now and tackle this remaining point on a different PR. I guess we'll talk about this on the upcoming sync, but leaving a comment for posterity!

Comment (lmbarros) [reply src/stage1.rs]: So, as agreed on our sync, I am marking this point as resolved and approving the PR! :rocket:

**Review by lmbarros — APPROVED** [review]
LGTM!
